web/modules/custom/dpl_graphql/src/EventSubscriber/CacheableResponseSubscriber.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace Drupal\dpl_graphql\EventSubscriber;
+
+use Drupal\Core\Cache\CacheableJsonResponse;
+use Drupal\Core\Session\AccountProxyInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * Add cache tags headers on cacheable responses, for external caching systems.
+ */
+class CacheableResponseSubscriber implements EventSubscriberInterface {
+
+  /**
+   * Constructor.
+   */
+  public function __construct(
+    protected AccountProxyInterface $currentUser,
+    protected LoggerInterface $logger,
+  ) {}
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    $events[KernelEvents::RESPONSE][] = ['onRespond'];
+    return $events;
+  }
+
+  /**
+   * Add cache tags headers on cacheable responses.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\ResponseEvent $event
+   *   The event to process.
+   */
+  public function onRespond(ResponseEvent $event): void {
+    if (!$event->isMainRequest()) {
+      return;
+    }
+
+    if (!$this->currentUser->hasPermission('get dpl graphql cache tags')) {
```

Comment:
We ought to find a better way to determine if it's a GraphQL request. Using a permission to enable GraphQL behaviour seems... Wrong.

Perhaps checking the current route if the controller is the `graphql` modules controller?